File: libschema/src/symbol/function.rs
use std::ops::ShlAssign;

use serde::{Deserialize, Serialize};

use crate::metadata::Metadata;
use crate::symbol::{Argument, Declaration, Metable};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Function {
    #[serde(flatten)]
    pub declaration: Declaration,

    /// Ex kinds: (forward, stock, etc)
    pub kind: String,

    /// Return type of the function
    pub return_type: String,

    /// Arguments of the function
    pub arguments: Vec<Argument>,
}

impl Metable for Function {
    fn metadata(&mut self) -> &mut Option<Metadata> {
        &mut self.declaration.documentation.metadata
    }
}

impl ShlAssign for Function {
    fn shl_assign(&mut self, rhs: Self) {
        self.declaration <<= rhs.declaration;
        self.kind = rhs.kind;
        self.return_type = rhs.return_type;
        self.arguments = rhs.arguments;
    }
}

File: typings/src/interfaces/symbol/function.ts
import { Declaration  } from './base';
import { Argument } from './argument';

export interface Function extends Declaration {
    /**
     * @brief Function kind (forward, stock, etc)
     * @readonly
     */
    readonly kind: FunctionKind,

    /**
     * @brief Return type of the function
     * @readonly
     */
    readonly returnType: string,

    /**
     * @brief Arguments of the function
     * @readonly
     */
    readonly arguments: Argument[],
}

export enum FunctionKind {
    Forward = 'forward',
    Native = 'native',
    Stock = 'stock',
    Static = 'static',
    StaticStock = 'static stock',
    Function = 'function',
}
